gateway/service/src/main/java/org/eclipse/ditto/gateway/service/endpoints/actors/CheckPermissionsActor.java

```diff
@@ -0,0 +1,417 @@
+     * <p>
+     * If the entity is a policy resource, it immediately creates a {@link PolicyId} from the entity ID.
+     * Otherwise, it retrieves the {@link PolicyId} using the Ditto Things API by sending a {@link SudoRetrieveThing} command.
+     * <p>
+     * If an unexpected response type is received a {@link DittoRuntimeException}
+     * or a {@link DittoInternalErrorException} will be thrown to indicate an internal processing error.
+     * <p>
+     * For cases where the response is a {@link DittoRuntimeException}, the exception is rethrown directly.
+     * If the response is not of the expected type or if the {@link PolicyId} is absent, an {@link IllegalArgumentException}
+     * wrapped inside a {@link DittoInternalErrorException} is thrown, ensuring proper error handling.
+     *
+     * @param entityId the ID of the entity for which the {@link PolicyId} is being retrieved.
+     * @param isPolicyResource flag indicating whether the entity is a policy resource.
+     * @param headers the {@link DittoHeaders} associated with the request.
+     * @return a {@link CompletionStage} that completes with the {@link PolicyId} once the operation is finished.
+     * @throws DittoRuntimeException if an unexpected response is received, or the {@link PolicyId} is missing in the response.
+     * @throws DittoInternalErrorException if an unexpected response type is encountered, or if the {@link PolicyId} is missing.
+     */
+    private CompletionStage<PolicyId> retrievePolicyIdForEntity(String entityId, boolean isPolicyResource,
+            DittoHeaders headers) {
+        if (isPolicyResource) {
+            return CompletableFuture.completedFuture(PolicyId.of(entityId));
+        } else {
+            final SudoRetrieveThing retrieveThing = SudoRetrieveThing.of(
+                    ThingId.of(entityId),
+                    JsonFieldSelector.newInstance(Thing.JsonFields.POLICY_ID.getPointer()),
+                    headers);
+
+            return Patterns.ask(edgeCommandForwarder, retrieveThing, defaultAskTimeout)
+                    .thenApply(result -> {
+                        if (result instanceof SudoRetrieveThingResponse thingResponse) {
+                            return thingResponse.getThing().getPolicyId().orElse(null);
+                        } else if (result instanceof DittoRuntimeException) {
+                            throw (DittoRuntimeException) result;
+                        } else {
+                            throw DittoInternalErrorException.newBuilder()
+                                    .dittoHeaders(headers)
+                                    .cause(new IllegalArgumentException(
+                                            "Unexpected response type: " + result.getClass().getSimpleName()))
+                                    .build();
+                        }
+                    }).handle((policyId, throwable) -> {
```

Comment:
You can combine the above `thenApply` and this `handle` to just a `handle` - then you don't have to create a `DittoInternalErrorException` twice and simplify the code..